netaddr.go

```diff
@@ -47,7 +41,22 @@ const o4 = len(mapped4Prefix)
 // Its memory representation is 24 bytes on 64-bit machines (the same
 // size as a Go slice header) for both IPv4 and IPv6 address.
 type IP struct {
-	a [16]byte // IPv6 or IPv4-mapped IPv6 form of IPv4 addr
+	// hi and lo are the bits of an IPv6 address. If z==z4, hi and lo
+	// contain the IPv4-mapped IPv6 address.
+	//
+	// hi and lo are constructed by interpreting a 16-byte IPv6
+	// address as a big-endian 128-bit number. The most significant
+	// bits of that number go into hi, the rest into lo.
+	//
+	// For example, 0011:2233:4455:6677:8899:aabb:ccdd:eeff is stored as:
+	//  hi = 0x0011223344556677
+	//  lo = 0x8899aabbccddeeff
+	//
+	// We store IPs like this, rather than as [16]byte, because it
```

Comment:
This change is okay, but the premise makes me a little sad.

How much of the 80% improvement do you get by just using the new Prefix code but keeping the old representation and doing the BigEndian.Uint64([:8]), BigEndian.Uint64([8:]) in the Prefix method first? Since the compiler *intrinsifies* those calls (_better_ than inlining), it should be pretty good.

Reply:
They’re not intrinsics. That’s part of how we ended up with broken compiler optimizations for them. See also https://github.com/golang/go/issues/42958, in which I suggest making them incrementally more intrinsic-like.

Reply:
Also worth noting is that the compiler can (theoretically) always do cheap binary tricks on amd64, which allows unaligned access. But without a way to specify alignment (such as https://github.com/golang/go/issues/19057 or https://github.com/golang/go/issues/6752), the compiler can’t be sure that the struct is aligned, arm systems can only benefit when we use uint64s.

Reply:
Cool, thanks for the links. Can you link all those (or at least link to this PR thread) in the commit message?

By default it seems difficult to map back from a commit back to its code review on GitHub.

Reply:
Added links to this PR, and to the Go bugs we've linked from here, along with an explanatory paragraph explaining "why not just change Prefix?"